specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts

```diff
@@ -234,65 +233,158 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
       ),
     };
 
-    const hasSameValues = (
+    const getFieldValue = async (
+      resource: SpecifyResource<AnySchema>,
+      fieldName: string
+    ): Promise<{
+      readonly id: number | undefined;
+      readonly cid: string | undefined;
+      readonly value: number | string | null | undefined;
+    }> => {
+      const localCollection = resource.collection ?? {
+        models: [],
+        field: undefined,
+      };
+
+      const collectionField = localCollection.field;
+      if (
+        localCollection.related !== undefined &&
+        collectionField?.name.toLowerCase() ===
+          fieldName.split(djangoLookupSeparator)[0].toLowerCase()
+      ) {
+        const related = localCollection.related;
+        const splitName = fieldName.split(djangoLookupSeparator);
+        return splitName.length === 1
+          ? ({ id: related.id, cid: related.cid, value: related.id } as const)
+          : getFieldValue(
+              related,
+              fieldName
+                .split(djangoLookupSeparator)
+                .slice(1)
+                .join(djangoLookupSeparator)
+            );
+      }
+
+      const related: SpecifyResource<AnySchema> | number | string | null =
+        await resource.getRelated(
+          fieldName.replaceAll(djangoLookupSeparator, backboneFieldSeparator)
+        );
+      return {
+        id: related?.id,
+        cid: related?.cid,
+        value:
+          related?.cid === undefined
+            ? (related as unknown as number | string | null)
+            : related.id,
+      } as const;
+    };
+
+    const generateFilters = async (
+      resource: SpecifyResource<AnySchema>,
+      fieldNames: RA<string>
+    ): Promise<
+      IR<{
+        readonly id: number | undefined;
+        readonly cid: string | undefined;
+        readonly value: number | string | null | undefined;
+      }>
```

Comment:
this type seems to be repeated several times in this file - want to create a `type` for it?